# Knowledge-based Recommender System (KBRS)

Questo notebook guida attraverso le prime fasi di pre-elaborazione e arricchimento dei dati per un sistema di raccomandazione basato sulla conoscenza. Il processo si articola in tre passaggi principali:

1.  **Caricamento Dati**: Carica i file `.csv` grezzi in DataFrame di Pandas.
2.  **Preparazione Catalogo**: Unisce i dati dei film e i link per creare un catalogo unico di film.
3.  **Arricchimento Dati**: Arricchisce il catalogo film con informazioni da Wikidata e DBpedia.
4.  **Unione Dati Finali**: Unisce le valutazioni degli utenti con il catalogo arricchito.

In [7]:
%pip install SPARQLWrapper
%pip install pandas
%pip install tqdm

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 565.1/565.1 kB 14.9 MB/s eta 0:00:0000:01


## 1. Caricamento dei Dati
Iniziamo caricando i dataset `movies.csv`, `links.csv` e `ratings.csv` utilizzando le funzioni dal modulo `src.data_manager`.

In [9]:
import pandas as pd
import sys
import os
from IPython.display import display

# Aggiusta il percorso per l'import del modulo `src`
project_root = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
if project_root not in sys.path:
    sys.path.append(project_root)

from src.data_manager import load_data, join_dataframes, enrich_movies, clean_partial_rows, drop_columns, normalize_movie_data_parallel, normalize_ratings_data, compress_kbrs_dataset

# Carica i dati grezzi dai file CSV
ratings_df, movies_df, links_df = load_data()

if movies_df is not None and links_df is not None and ratings_df is not None:
    print("\n--- Dati caricati ---")

    print("movies.csv:", movies_df.shape)
    print("\nlinks.csv:", links_df.shape)
    print("\nratings.csv:", ratings_df.shape)
    display(movies_df.head())
    display(links_df.head())
    display(ratings_df.head())


Data loaded successfully.

--- Dati caricati ---
movies.csv: (9742, 3)

links.csv: (9742, 3)

ratings.csv: (43084, 4)


,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


,userId,movieId,rating,timestamp
0,1,1,4.0,964982703.0
1,1,3,4.0,964981247.0
2,1,6,4.0,964982224.0
3,1,47,5.0,964983815.0
4,1,50,5.0,964982931.0


## 2. Preparazione del Catalogo Film Unico
Uniamo i DataFrame `movies` e `links` per creare un catalogo di film unici. Questo passaggio è cruciale per evitare chiamate API duplicate e rendere il processo di arricchimento efficiente.

In [10]:
unique_movie_catalog = join_dataframes(movies_df, links_df)
unique_movie_catalog = unique_movie_catalog.drop(columns=['tmdbId'])
unique_movie_catalog['imdbId'] = unique_movie_catalog['imdbId'].apply(lambda x: f"tt{int(x):07d}")

print("Catalogo film unico creato.")
print("Shape del catalogo unico:", unique_movie_catalog.shape)
display(unique_movie_catalog.head())

Catalogo film unico creato.
Shape del catalogo unico: (9742, 4)


,movieId,title,genres,imdbId
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,tt0114709
1,2,Jumanji (1995),Adventure|Children|Fantasy,tt0113497
2,3,Grumpier Old Men (1995),Comedy|Romance,tt0113228
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,tt0114885
4,5,Father of the Bride Part II (1995),Comedy,tt0113041


## 3. Arricchimento del Catalogo Film
Ora arricchiamo il catalogo unico di film con i dati semantici di Wikidata e DBpedia. La funzione `enrich_movies` gestirà le chiamate API in batch e mostrerà l'avanzamento con una barra di caricamento.

In [11]:
enriched_movie_catalog = enrich_movies(unique_movie_catalog)

print("\n--- Dati arricchiti ---")
print("Shape del catalogo arricchito:", enriched_movie_catalog.shape)
enriched_movie_catalog = drop_columns(enriched_movie_catalog, ['title', 'imdbId', 'dbpediaAbstract', 'wikidataId'])
display(enriched_movie_catalog.head())

No checkpoint found. Starting from scratch.


Enriching movies: 100%|██████████| 9742/9742 [43:13<00:00,  3.76it/s]

Movie enrichment completed.

--- Dati arricchiti ---
Shape del catalogo arricchito: (9742, 9)


,movieId,genres,dbpediaDirector,dbpediaRuntime,dbpediaActors
0,1,Adventure|Animation|Children|Comedy|Fantasy,John Lasseter,4860.0,John Ratzenberger|Annie Potts|Tom Hanks|Erik v...
1,2,Adventure|Children|Fantasy,Joe Johnston,6240.0,Robin Williams|Bebe Neuwirth|Bonnie Hunt|David...
2,3,Comedy|Romance,Howard Deutch,6060.0,Ann-Margret|Walter Matthau|Daryl Hannah|Jack L...
3,4,Comedy|Drama|Romance,Forest Whitaker,7440.0,Angela Bassett|Lela Rochon|Loretta Devine|Whit...
4,5,Comedy,Charles Shyer,6360.0,George Newbern|Steve Martin|Eugene Levy|Kieran...


## 4. Pulizia dati vuoti


In [12]:
cleaned_df = None

try:
    # Call the cleaning function
    cleaned_df = clean_partial_rows(enriched_movie_catalog)
except Exception as e:
    print(f"An unexpected error occurred: {e}")


--- Cleaning Report ---
Total records before cleaning: 9742
Total records after cleaning: 8024
Deleted records: 1718

Empty records per column (before cleaning):
movieId               0
genres                0
dbpediaDirector    1101
dbpediaRuntime      673
dbpediaActors      1258

Cleaned data has been saved to 'data/processed/dbpedia_data_cleaned.csv'.


## 5. Normalizzazione dei dati
In questo passaggio le colonne numeriche vengono normalizzate da 0 a 1 e quelle testuali in one-hot encoding.

In [13]:
normalized_data= None

if cleaned_df is not None:
    normalized_data = normalize_movie_data_parallel(cleaned_df)
    print(normalized_data.shape)

    normalized_ratings = normalize_ratings_data(ratings_df)


Fase 1/3: Raccolta e selezione delle feature uniche...


Raccolta Metadati: 100%|██████████| 8024/8024 [00:00<00:00, 15201.86it/s]

Fase 2/3: Normalizzazione dei dati numerici...
Fase 3/3: Elaborazione in parallelo con 2 core...



Elaborazione in Parallelo: 100%|██████████| 41/41 [00:00<00:00, 251045.93it/s]


Elaborazione completata. Risultato salvato in data/processed/normalized_movies_optimized.csv
(8024, 16112)
Fase 1/2: Normalizzazione del punteggio...
Fase 2/2: Creazione di una matrice di interazione sparsa (User-Item)...


In [14]:
    display(normalized_data.head())
    display(normalized_ratings.head())


,movieId,genre_(no genres listed),genre_Action,genre_Adventure,genre_Animation,genre_Children,genre_Comedy,genre_Crime,genre_Documentary,genre_Drama,...,actor_Émilien Néron,actor_Éric Berger,actor_Éric Judor,actor_Érica Rivas,actor_Óscar Jaenada,actor_Özge Özberk,actor_Đơn Dương,actor_Şerif Sezer,actor_Željko Ivanek,runtime_normalized
0,1,0,0,1,1,1,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0.358170
1,2,0,0,1,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0.388235
2,3,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0.384314
3,4,0,0,0,0,0,1,0,0,1,...,0,0,0,0,0,0,0,0,0,0.414379
4,5,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0.390850


movieId,1,2,3,4,5,6,7,8,9,10,...,191005,193565,193567,193571,193573,193579,193581,193583,193585,193587
userId,,,,,,,,,,,,,,,,,,,,,
1,0.777778,0.0,0.777778,0.0,0.0,0.777778,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.000000,0.0,0.000000,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,0.000000,0.0,0.000000,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.000000,0.0,0.000000,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
5,0.777778,0.0,0.000000,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


## 6. Reduce data dimension
Application of TrunctedSVD to get a lighter representation of the data.

In [15]:
compressed_df = None

if normalized_data is not None:
    compressed_df = compress_kbrs_dataset(normalized_data, 4096)
    compressed_df.to_csv("data/processed/compressed_movie_catalog.csv", index=False)


Fase 1/3: Preparazione dei dati...
Dataset originale: 8024 righe, 16111 colonne.
Dimensioni occupate in memoria (stimato, solo features): 0.60 MB
Fase 2/3: Applicazione di TruncatedSVD con 4096 componenti...


Calcolo SVD: 100%|██████████| 1/1 [04:42<00:00, 282.54s/it]


Fase 3/3: Creazione del nuovo DataFrame compresso...
Dataset compresso: 8024 righe, 4097 colonne.
Varianza spiegata: 0.9214


In [16]:
print(compressed_df.shape)

(8024, 4097)


# Calcolo della Matrice di Similarità (TODO 1)

In [17]:
from sklearn.metrics.pairwise import cosine_similarity

# Isola le 4096 colonne di feature
# Assuming 'movieId' is the first column, we select all columns except the first one.
features_df = compressed_df.iloc[:, 1:]

# Calcola la matrice di similarità del coseno
cosine_sim_matrix = cosine_similarity(features_df)

print("Matrice di similarità del coseno calcolata.")
print("Shape della matrice di similarità:", cosine_sim_matrix.shape)

Matrice di similarità del coseno calcolata.
Shape della matrice di similarità: (8024, 8024)


## TODO
Una volta che hai il tuo dataset compresso (8024 film x 4097 colonne), sei pronto per procedere alla creazione del tuo **sistema di raccomandazione basato sulla conoscenza (KBRS)**. Nonostante il processo possa sembrare complesso, si basa su principi semplici: la **similarità tra gli elementi**.

---

### 1. Calcolo della Matrice di Similarità

La prima e più importante fase è calcolare quanto ogni film è simile a ogni altro film. La tua matrice compressa è perfetta per questo. Poiché le feature sono ora vettori densi e numerici, puoi usare la **similarità del coseno** per misurare l'angolo tra i vettori di due film. Un angolo piccolo indica un'alta similarità.

* **Perché la similarità del coseno?** La similarità del coseno è la metrica ideale per gli embedding. A differenza della distanza euclidea, non è influenzata dalla magnitudine dei vettori, ma solo dalla loro direzione. In questo caso, la direzione rappresenta le "caratteristiche latenti" del film.

**Passaggi:**

1.  Isola le 4096 colonne di feature dal DataFrame compresso.
2.  Utilizza la funzione `cosine_similarity` di scikit-learn per calcolare la matrice di similarità tra tutti i film.
3.  Il risultato sarà una matrice quadrata (8024x8024), dove ogni cella `(i, j)` conterrà il valore di similarità tra il film `i` e il film `j`.

---

### 2. Sviluppo della Logica di Raccomandazione

Con la matrice di similarità a disposizione, puoi costruire un motore di raccomandazione. La logica più comune è il **filtraggio basato sugli item (Item-Based Filtering)**.

**Passaggi:**

1.  **Input:** Il sistema riceve l'ID di un film che l'utente ha apprezzato (es. ha dato un voto alto).
2.  **Ricerca:** La matrice di similarità viene consultata per trovare i film più simili a quello di input.
3.  **Output:** Il sistema restituisce una lista dei film con il più alto punteggio di similarità.

Questa logica è semplice ed efficace, ma ha una limitazione: non tiene conto delle preferenze individuali dell'utente. Per renderla più potente, puoi combinarla con i dati di rating, come discusso in precedenza.

---

### 3. Integrazione con i Dati Utente (KBRS Ibrido)

Per creare un KBRS ibrido, che è molto più potente, devi combinare la matrice di similarità dei film con le preferenze degli utenti.

**Passaggi:**

1.  **Profilo Utente:** Costruisci un "profilo" per ogni utente, che sia un vettore delle sue preferenze. Per farlo, puoi prendere il rating medio di un utente per i film che ha visto e usarlo come un "peso" per le similarità.
2.  **Predizione del Rating:** Per ogni film che l'utente non ha ancora visto, il sistema può calcolare un "rating previsto". Questo si fa prendendo la somma delle similarità tra quel film e tutti i film che l'utente ha già apprezzato, pesata per i rating che l'utente ha dato a quei film.
3.  **Raccomandazione:** I film con il rating previsto più alto vengono raccomandati all'utente.

Questo approccio ibrido sfrutta il meglio di entrambi i mondi: la conoscenza profonda dei film (KBRS) e il comportamento degli utenti (filtraggio collaborativo).

In [34]:
import numpy as np

def create_user_profile(user_id, ratings_df, cleaned_df):
    """
    Crea un profilo utente basato sui rating forniti dall'utente.

    Args:
        user_id (int): L'ID dell'utente.
        ratings_df (pd.DataFrame): DataFrame con i dati di rating.
        cleaned_df (pd.DataFrame): DataFrame con il catalogo unico dei film.

    Returns:
        pd.DataFrame: DataFrame con i film visti dall'utente e i loro rating.
    """
    user_ratings = ratings_df[ratings_df['userId'] == user_id].copy()
    user_profile = pd.merge(user_ratings, unique_movie_catalog[['movieId', 'title']], on='movieId')
    return user_profile

def predict_rating(user_id, movie_id, ratings_df, cosine_sim_matrix, movie_ids):
    """
    Prevede il rating di un utente per un film non ancora visto,
    basato sulla similarità del coseno e i rating dell'utente.

    Args:
        user_id (int): L'ID dell'utente.
        movie_id (int): L'ID del film per cui prevedere il rating.
        ratings_df (pd.DataFrame): DataFrame con i dati di rating.
        cosine_sim_matrix (np.ndarray): La matrice di similarità del coseno.
        movie_ids (pd.Series): Una Series contenente gli ID dei film nell'ordine della matrice di similarità.

    Returns:
        float: Il rating previsto per il film, o None se non ci sono film visti con similarità.
    """
    user_ratings = ratings_df[ratings_df['userId'] == user_id]

    # Filtra i film che l'utente ha visto
    seen_movies = user_ratings['movieId'].tolist()

    # Ottieni l'indice del film per cui si vuole prevedere il rating
    try:
        target_movie_index = movie_ids[movie_ids == movie_id].index[0]
    except IndexError:
        print(f"Film con ID {movie_id} non trovato nel catalogo compresso.")
        return None

    # Inizializza variabili per il calcolo del rating previsto
    weighted_sum = 0
    similarity_sum = 0

    # Itera sui film visti dall'utente
    for seen_movie_id, rating in user_ratings[['movieId', 'rating']].values:
        try:
            seen_movie_index = movie_ids[movie_ids == seen_movie_id].index[0]
        except IndexError:
            # Salta i film visti che non sono nel catalogo compresso (dovrebbe essere raro)
            continue

        # Ottieni la similarità tra il film target e il film visto
        similarity = cosine_sim_matrix[target_movie_index, seen_movie_index]

        # Pesa la similarità con il rating dell'utente
        weighted_sum += similarity * rating
        similarity_sum += abs(similarity) # Usa il valore assoluto per evitare similarità negative

    # Calcola il rating previsto
    if similarity_sum > 0:
        predicted_rating = weighted_sum / similarity_sum
        return predicted_rating
    else:
        # Nessun film visto dall'utente ha similarità con il film target
        return None

import numpy as np

def recommend_movies_hybrid_optimized(user_id, ratings_df, cosine_sim_matrix, movie_ids, cleaned_df, num_recommendations=10):
    """
    Genera raccomandazioni per un utente utilizzando un approccio ibrido ottimizzato.

    Args:
        user_id (int): L'ID dell'utente.
        ratings_df (pd.DataFrame): DataFrame con i dati di rating.
        cosine_sim_matrix (np.ndarray): La matrice di similarità del coseno.
        movie_ids (pd.Series): Una Series contenente gli ID dei film nell'ordine della matrice di similarità.
        cleaned_df (pd.DataFrame): DataFrame con il catalogo unico dei film.
        num_recommendations (int): Il numero di raccomandazioni da generare.

    Returns:
        list: Una lista di tuple (movie_id, predicted_rating) dei film raccomandati.
    """
    user_ratings = ratings_df[ratings_df['userId'] == user_id]
    seen_movies_ids = user_ratings['movieId'].tolist()

    # Initialize dictionary to store aggregated scores for unseen movies
    aggregated_scores = {}
    similarity_sums = {} # To normalize the scores later

    # Get all movie IDs from the cleaned catalog
    all_movie_ids = cleaned_df['movieId'].tolist()

    # Iterate over the movies the user has rated
    for index, row in user_ratings.iterrows():
        seen_movie_id = row['movieId']
        rating = row['rating']

        try:
            # Get the index of the seen movie in the similarity matrix
            seen_movie_index = movie_ids[movie_ids == seen_movie_id].index[0]
        except IndexError:
            # Skip if the seen movie is not in the compressed catalog
            continue

        # Get similarity scores for the seen movie with all other movies
        similarity_scores = cosine_sim_matrix[seen_movie_index]

        # Iterate through all movies to aggregate scores
        for i, movie_id in enumerate(movie_ids):
            # Skip movies the user has already seen
            if movie_id in seen_movies_ids:
                continue

            similarity = similarity_scores[i]

            # Aggregate the weighted score
            if movie_id not in aggregated_scores:
                aggregated_scores[movie_id] = 0
                similarity_sums[movie_id] = 0

            aggregated_scores[movie_id] += similarity * rating
            similarity_sums[movie_id] += abs(similarity) # Use absolute similarity for normalization

    # Normalize the aggregated scores to get predicted ratings
    predicted_ratings = []
    for movie_id, score_sum in aggregated_scores.items():
        if similarity_sums[movie_id] > 0:
            predicted_rating = score_sum / similarity_sums[movie_id]
            predicted_ratings.append((movie_id, predicted_rating))

    # Sort movies by predicted rating in descending order
    predicted_ratings = sorted(predicted_ratings, key=lambda x: x[1], reverse=True)

    # Return the top recommendations
    return predicted_ratings[:num_recommendations]

 #Example of usage (can be uncommented and run in a separate cell after defining the function)
example_user_id = 1
movie_ids_series = compressed_df['movieId'] # Assuming 'compressed_df' is available
recommended_movies_optimized = recommend_movies_hybrid_optimized(example_user_id, ratings_df, cosine_sim_matrix, movie_ids_series, cleaned_df)

print(f"\nRaccomandazioni ottimizzate per l'utente {example_user_id}:")
for movie_id, predicted_rating in recommended_movies_optimized:
    movie_title = unique_movie_catalog[unique_movie_catalog['movieId'] == movie_id]['title'].iloc[0]
    print(f"- Movie ID: {movie_id}, Titolo: {movie_title}, Rating previsto: {predicted_rating:.4f}")


Raccomandazioni ottimizzate per l'utente 1:
- Movie ID: 31193, Titolo: Many Adventures of Winnie the Pooh, The (1977), Rating previsto: 4.6137
- Movie ID: 85259, Titolo: Winnie the Pooh and the Honey Tree (1966), Rating previsto: 4.6036
- Movie ID: 60803, Titolo: Little Drummer Boy, The (1968), Rating previsto: 4.5841
- Movie ID: 3725, Titolo: American Pop (1981), Rating previsto: 4.5801
- Movie ID: 85316, Titolo: Winnie the Pooh and Tigger Too (1974), Rating previsto: 4.5778
- Movie ID: 91488, Titolo: Snowman, The (1982), Rating previsto: 4.5746
- Movie ID: 1489, Titolo: Cats Don't Dance (1997), Rating previsto: 4.5722
- Movie ID: 5629, Titolo: Jonah: A VeggieTales Movie (2002), Rating previsto: 4.5700
- Movie ID: 3159, Titolo: Fantasia 2000 (1999), Rating previsto: 4.5698
- Movie ID: 3213, Titolo: Batman: Mask of the Phantasm (1993), Rating previsto: 4.5619
